datamol/fp.py

```diff
@@ -3,59 +3,152 @@
+def fp_to_array(
+    fp: Union[np.ndarray, SparseBitVect, ExplicitBitVect, UIntSparseIntVect]
+) -> np.ndarray:
     """Convert rdkit fingerprint to numpy array.
 
     Note:
         This implementation has shown to be faster than using `DataStructs.ConvertToNumpyArray`
-        by a factor of ~4.
+        by a factor of ~4. See https://github.com/rdkit/rdkit/discussions/3863.
+
+    Args:
+        fp: The fingerprint.
     """
+
     if isinstance(fp, np.ndarray):
-        return fp
-    return np.frombuffer(fp.ToBitString().encode(), "u1") - ord("0")
+        fp_out = fp
+
+    elif isinstance(fp, SparseBitVect):
+        tmp = np.zeros(fp.GetNumBits(), dtype=int)
+        for n_bit in list(fp.GetOnBits()):
+            tmp[n_bit] = 1
+        fp_out = tmp
+
+    elif isinstance(fp, ExplicitBitVect):
+        fp_out = np.frombuffer(fp.ToBitString().encode(), "u1") - ord("0")
+
+    elif isinstance(
+        fp, (UIntSparseIntVect, IntSparseIntVect, LongSparseIntVect, ULongSparseIntVect)
+    ):
+        # one of the other rdkit type
+        tmp = np.array(0, dtype="int")
+        ConvertToNumpyArray(fp, tmp)
+        fp_out = tmp
+
+    else:
+        raise ValueError(
+            f"The fingerprint of type '{type(fp)}' is not supported. "
+            "Please open a ticket at https://github.com/datamol-org/datamol/issues."
+        )
+
+    return fp_out
 
 
 def to_fp(
     mol: Union[str, Chem.rdchem.Mol],
     fp_size: int = 2048,
     radius: int = 3,
     use_features: bool = True,
+    fp_args: dict = None,
     as_array: bool = True,
-) -> Optional[Union[np.ndarray, DataStructs.ExplicitBitVect]]:
-    """Transform a molecule from smiles to morgan fingerprint.
-
-    Note:
-        That function should be expanded to compute more type of fingerprints.
+    fp_type: str = "ecfp",
+) -> Optional[Union[np.ndarray, SparseBitVect, ExplicitBitVect, UIntSparseIntVect]]:
+    """Compute the molecular fingerprint given a molecule or a SMILES.
 
     Args:
-        mol (Chem.Mol or str): a molecule or a SMILES.
-        fp_size (int, optional): Size of morgan fingerprint. Default to 2048.
-        radius (int, optional): Radius of the morgan fingerprints. Default to 3.
+        mol: a molecule or a SMILES.
+        fp_size: Size of morgan fingerprint. Default to 2048.
+            Warning: only apply to the default ECFP fingerprint.
+        radius: Radius of the morgan fingerprints. Default to 3.
+            Warning: only apply to the default ECFP fingerprint.
         use_features: Whether to use atom features. Default to True.
+            Warning: only apply to the default ECFP fingerprint.
+        fp_args: Arguments to build the fingerprint. Refer to the official RDKit documentation.
         as_array: Whether to return a numpy array of an RDKit vec. Default to True.
+        fp_type: The type of the fingerprint. See `dm.list_supported_fingerprints()` for a
+            complete list.
```

Comment:
Instead of `fp_args` maybe I should just use `**kwargs` here and throw a warning whenever `use_features` and `fp_size` is used instead of `useFeatures` and `nBits`.

Reply:
New API is better now and backward compatible (with warnings).